### PR TITLE
Remove hot string allocations in MessagingStatisticsGroup

### PR DESCRIPTION
**Pull request by ReubenBond**
Avoids ~900 bytes of allocs per call.

Before:

 Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
------- |---------:|---------:|---------:|-------:|----------:|
   Ping | 338.1 us | 254.3 us | 14.37 us | 3.4180 |  15.83 KB |

After:

 Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
------- |---------:|---------:|---------:|-------:|----------:|
   Ping | 326.5 us | 85.51 us | 4.832 us | 3.4180 |  14.93 KB |

### CODE REVIEW

**Comment by ReubenBond** [comment]
@dotnet-bot test functional

**Comment by ReubenBond** [comment]
What's going on here is that instead of constructing a complex string once for every call and using that to index a dictionary, we're using the parameter used to construct that string to index the dictionary.

File: src/Orleans.Core/Statistics/MessagingStatisticsGroup.cs [hunk 1]
```diff
@@ -56,13 +56,31 @@ internal void OnMessage(int numMsgsInBatch, int totalBytes)
...
+            (new StatisticName(StatisticNames.MESSAGING_RECEIVED_MESSAGES_PER_SILO, addr != null ? addr.ToString() : "Null"), CounterStorage.LogOnly);
+
+        private static readonly Func<SiloAddress, (StatisticName, CounterStorage)> GetPingSendCounterDetails =
+            destination => (new StatisticName(StatisticNames.MESSAGING_PINGS_SENT_PER_SILO, destination.ToString()), CounterStorage.LogOnly);
+
+        private static readonly Func<SiloAddress, (StatisticName, CounterStorage)> GetPingReceiveCounterDetails = destination =>
+            (new StatisticName(StatisticNames.MESSAGING_PINGS_RECEIVED_PER_SILO, destination.ToString()), CounterStorage.LogOnly);
+
+        private static readonly Func<SiloAddress, (StatisticName, CounterStorage)> GetPingReplyReceivedCounterDetails = replier =>
+            (new StatisticName(StatisticNames.MESSAGING_PINGS_REPLYRECEIVED_PER_SILO, replier.ToString()), CounterStorage.LogOnly);
```
Comment (dVakulen) [inline src/Orleans.Core/Statistics/MessagingStatisticsGroup.cs]: Nit: underscore between `REPLY` and `RECEIVED` might help readability.- `

**Comment by jdom** [comment]
What happened?

**Comment by ReubenBond** [comment]
We'll bring these issues back when there's more time to review them

**Pull request closed without merging**